### PR TITLE
feat(module/virtual_network_gateway): New module for Site-to-Site VPN connections

### PR DESCRIPTION
**Pull request by sebastianczech**
## Description

PR delivers new module for Site-to-Site VPN connections, which internally:

- creates Azure Virtual Network Gateway
- creates Azure Local Network Gateway (which in fact is customer gateway)
- creates VNG connections

## Motivation and Context

#303 

## How Has This Been Tested?

Code was tested by deploying VM-Series in common design.
In GCP cloud it was created another environment with Panorama .
Then there were established tunnels between Azure (using that module) and GCP and VM-Series from Azure were bootstrapped by Panorama in GCP.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/virtual_network_gateway/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,7 @@
+output "public_ip" {
+  description = "Public IP addresses for Virtual Network Gateway"
+  value = merge(
+    { for k, v in azurerm_public_ip.this : k => v.ip_address },
+    { for k, v in data.azurerm_public_ip.exists : k => v.ip_address }
+  )
+}
```
[the remark this answers is not in the document]
Comment (sebastianczech) [reply modules/virtual_network_gateway/outputs.tf:7]: I've added additional output with IPsec policy. Thank you for feedback.

**Review by migara — APPROVED** [review]

**Comment by sebastianczech** [comment]
@FoSix according to our offline discussion, I've updated PR about:
- adding `azure_bgp_peers_addresses`, which is a map of addresses used on Azure side (then in other parts of configuration we are referencing key)
- introducing map for `peering_addresses` in `local_bgp_settings` (instead of list), in which also we are referencing to keys for IP address
- merging changes from `main` branch e.g. removing `Makefile` and adding Terratest

File: modules/virtual_network_gateway/variables.tf [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+variable "azure_bgp_peers_addresses" {
+  description = <<-EOF
+  Map of IP addresses used on Azure side for BGP. Map is used to not to duplicate IP address and refer to keys while configuring:
+  - custom_bgp_addresses
+  - peering_addresses in local_bgp_settings
+
+  Example:
+
+  azure_bgp_peers_addresses = {
```
Comment (FoSix) [inline modules/virtual_network_gateway/variables.tf:153]: same suggestion as [here](https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/pull/306#discussion_r1321159362)

File: modules/virtual_network_gateway/variables.tf [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+  type        = string
+}
+
+variable "vpn_type" {
+  description = "The routing type of the Virtual Network Gateway. Valid options are RouteBased or PolicyBased. Defaults to RouteBased. Changing this forces a new resource to be created."
+  default     = "RouteBased"
+  type        = string
+}
+
+variable "sku" {
```
Comment (FoSix) [inline modules/virtual_network_gateway/variables.tf:60]: This depends on Generation and SKU, should we add some validation? Or just rely on error messages thrown by the API?

**Review by FoSix — COMMENTED** [review]
:+1:
just added some small, optional ideas

File: modules/virtual_network_gateway/variables.tf [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+  - peering_addresses - a map of peering addresses, which contains 1 (for active-standby) or 2 objects (for active-active) with:
+    - key is the ip configuration name
+    - apipa_addresses is the list of keys for IP addresses defined in variable azure_bgp_peers_addresses
+  - peer_weight - the weight added to routes which have been learned through BGP peering. Valid values can be between 0 and 100.
+
+  Example:
+
+  local_bgp_settings = {
+    asn = "65001"
+    peering_addresses = {
```
[the remark this answers is not in the document]
Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:177]: Unfortunately in Terraform the condition for `validation` in variable can only refer to the variable itself, using , so we cannot for `local_bgp_settings` refer to value of `active_active`, in which we are defining type of VNG (active-active or single node). 

In variable description I've already put that information: https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/cfd34e1c6e0d0e4038d4a8632570f6aa03622b7e/modules/virtual_network_gateway/variables.tf#L168-L170

Could it be sufficient ?

File: modules/virtual_network_gateway/variables.tf (hunk 2 above)
Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:153]: same reply as [here](https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/pull/306#discussion_r1321333921)

File: modules/virtual_network_gateway/variables.tf (hunk 3 above)
Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:68]: Thank you for feedback, you are right.

I've added validation for `type`, `vpn_type`, `sku` and `generation` e.g.: https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/64b2fbfce7645aa94d04ec23e2c2a11f8be6e2b1/modules/virtual_network_gateway/variables.tf#L99-L107

File: modules/virtual_network_gateway/variables.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (FoSix) [reply modules/virtual_network_gateway/variables.tf:194]: no :) we could use `lifecycle { precondition {} }`. You can operate there on all available variables

File: modules/virtual_network_gateway/variables.tf (hunk 3 above)
Comment (FoSix) [reply modules/virtual_network_gateway/variables.tf:68]: i was not clear enough, I thought about adding precondition block

Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:68]: @FoSix , no problem, thank you for feedback.

I've added `lifecycle precondition` to check SKU, which depends on generation: https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/872dce18522fce79418298c412c036ccf8e34c7b/modules/virtual_network_gateway/main.tf#L80C3-L85

I hope it fullfil requirement about check generation and SKU.

File: modules/virtual_network_gateway/variables.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:194]: @FoSix thank you for feedback.

I've added checking number of peering addresses:

https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/872dce18522fce79418298c412c036ccf8e34c7b/modules/virtual_network_gateway/main.tf#L86-L89

File: modules/virtual_network_gateway/variables.tf (hunk 2 above)
Comment (sebastianczech) [reply modules/virtual_network_gateway/variables.tf:170]: I've added checking number of Azure BGP APIPA IP addresses:

https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/872dce18522fce79418298c412c036ccf8e34c7b/modules/virtual_network_gateway/main.tf#L90-L93

**Review by FoSix — APPROVED** [review]

**Pull request merged**